pypif_sdk/func/calculate_funcs.py

```diff
@@ -0,0 +1,258 @@
+
+def calculate_ideal_atomic_percent(pif):
+    """
+    Calculates ideal atomic percents from a chemical formula string from a pif. Returns an appended pif with composition elements modified or added.
+    """
+    try:
+        pif.chemicalFormula
+    except NameError:
+        return pif
+    else:
+        expanded_formula = __expand_formula__(pif.chemicalFormula)
+        e_array = __create_emprical_compositional_array__(expanded_formula)
+        appended_e_array = __add_atomic_percents__(e_array)
+        for e in appended_e_array:
+            if __check_if_element_exists_in_pif_composition__(pif, e["symbol"]):
+                in_pif = __get_element_in_pif_composition__(pif, e["symbol"])
+                comp = in_pif[0]
+                pif.composition.pop(in_pif[1])
+                comp.idealAtomicPercent = e["atomic_percent"]
+                pif.composition.append(comp)
+            else:
+                comp = Composition()
+                comp.element = e["symbol"]
+                comp.idealAtomicPercent = e["atomic_percent"]
+                pif.composition.append(comp)
+        return pif
+
+def calculate_ideal_weight_percent(pif):
+    """
+    Calculates ideal atomic weight percents from a chemical formula string from a pif. Returns an appended pif with composition elements modified or added.
+    """
+    try:
+        pif.chemicalFormula
+    except NameError:
+        return pif
+    else:
+        expanded_formula = __expand_formula__(pif.chemicalFormula)
+        e_array = __create_emprical_compositional_array__(expanded_formula)
+        appended_e_array = __add_ideal_atomic_weights__(e_array)
+        a_array_with_pcts = __add_ideal_weight_percent__(appended_e_array)
+        for e in a_array_with_pcts:
+            if __check_if_element_exists_in_pif_composition__(pif, e["symbol"]):
+                in_pif = __get_element_in_pif_composition__(pif, e["symbol"])
+                comp = in_pif[0]
+                pif.composition.pop(in_pif[1])
+                comp.idealWeightPercent = e["weight_percent"]
+                pif.composition.append(comp)
+            else:
+                comp = Composition()
+                comp.element = e["symbol"]
+                comp.idealWeightPercent = e["weight_percent"]
+                pif.composition.append(comp)
+        return pif
+
+def __expand_formula__(formula_string):
+    """
+    Accounts for the many ways a user may write a formula string, and returns an expanded chemical formula string.
+    Assumptions:
+    -The Chemical Formula string it is supplied is well-written, and has no hanging parethneses
+    -The number of repeats occurs after the elemental symbol or ) ] character EXCEPT in the case of a hydrate where it is assumed to be in front of the first element
+    -All hydrates explicitly use the · symbol
+    -Only (, (,[, ], ., · are "important" symbols to intrepreting the string.
+    -IONS ARE NOT HANDLED
+
+    Takes in a messy chemical formula string
```

Comment:
The other methods use `:param :` annotations; any reason not to include them here too?